In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
%matplotlib inline 

file1= 'df_final_demo.txt'
file2='df_final_experiment_clients.txt'
file3='df_final_web_data_pt_1.txt'
file4='df_final_web_data_pt_2.txt'
df1 = pd.read_csv(file4)
df2 = pd.read_csv(file3)
df3 = pd.read_csv(file2)
df4 = pd.read_csv(file1)

In [2]:
df1 = pd.read_csv(file3)
df1

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
343136    2443347  465784886_73090545671  136329900_10529659391_316129   
343137    2443347  465784886_73090545671  136329900_10529659391_316129   
343138    2443347  465784886_73090545671  136329900_10529659391_316129   
343139    2443347  465784886_73090545671  136329900_10529659391_316129   
343140    2443347  465784886_73090545671  136329900_10529659391_316129   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
343136      confirm  2017-03-31 15:15:46  
343137       step_3  2017-03-31 15:14:53  
343138       step_2  2017-03-31 15:12:08  
343139       step_1  2017-03-31 15:11:37  
343140        start  2017-03-31 15:10:47  

[343141 rows x 5 columns]

In [3]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343141 entries, 0 to 343140
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     343141 non-null  int64 
 1   visitor_id    343141 non-null  object
 2   visit_id      343141 non-null  object
 3   process_step  343141 non-null  object
 4   date_time     343141 non-null  object
dtypes: int64(1), object(4)
memory usage: 13.1+ MB


In [4]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343141 entries, 0 to 343140
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     343141 non-null  int64 
 1   visitor_id    343141 non-null  object
 2   visit_id      343141 non-null  object
 3   process_step  343141 non-null  object
 4   date_time     343141 non-null  object
dtypes: int64(1), object(4)
memory usage: 13.1+ MB


In [5]:
df1.nunique()

client_id        58391
visitor_id       62936
visit_id         75256
process_step         5
date_time       282666
dtype: int64

In [6]:
df2.nunique()

client_id        58391
visitor_id       62936
visit_id         75256
process_step         5
date_time       282666
dtype: int64

In [7]:
# Display duplicated rows 
df1.loc[df1.duplicated(keep=False)]

client_id             visitor_id                      visit_id  \
364       9358626  148475529_63668223960  223297395_36250329195_832161   
365       9358626  148475529_63668223960  223297395_36250329195_832161   
367       5376894  622509328_44408201172  688984457_43441834354_912755   
368       5376894  622509328_44408201172  688984457_43441834354_912755   
382        438567  923318467_50708963627  330543236_29863358529_771432   
...           ...                    ...                           ...   
342679    1251269  791822615_78146852845   763845944_55757914723_70888   
342684    8350388  304440669_31040182888  143359398_96671536668_299057   
342685    8350388  304440669_31040182888  143359398_96671536668_299057   
342944    4553664  795008403_33653570753  439592710_65772985411_546352   
342945    4553664  795008403_33653570753  439592710_65772985411_546352   

       process_step            date_time  
364           start  2017-04-28 12:27:28  
365           start  2017-04-28 12:27:28  
367           start  2017-04-28 14:20:01  
368           start  2017-04-28 14:20:01  
382           start  2017-04-28 02:52:29  
...             ...                  ...  
342679        start  2017-04-28 10:18:20  
342684        start  2017-04-28 12:44:21  
342685        start  2017-04-28 12:44:21  
342944        start  2017-04-28 07:07:17  
342945        start  2017-04-28 07:07:17  

[4124 rows x 5 columns]

In [8]:
# Display duplicated rows 
df2.loc[df2.duplicated(keep=False)]

client_id             visitor_id                      visit_id  \
364       9358626  148475529_63668223960  223297395_36250329195_832161   
365       9358626  148475529_63668223960  223297395_36250329195_832161   
367       5376894  622509328_44408201172  688984457_43441834354_912755   
368       5376894  622509328_44408201172  688984457_43441834354_912755   
382        438567  923318467_50708963627  330543236_29863358529_771432   
...           ...                    ...                           ...   
342679    1251269  791822615_78146852845   763845944_55757914723_70888   
342684    8350388  304440669_31040182888  143359398_96671536668_299057   
342685    8350388  304440669_31040182888  143359398_96671536668_299057   
342944    4553664  795008403_33653570753  439592710_65772985411_546352   
342945    4553664  795008403_33653570753  439592710_65772985411_546352   

       process_step            date_time  
364           start  2017-04-28 12:27:28  
365           start  2017-04-28 12:27:28  
367           start  2017-04-28 14:20:01  
368           start  2017-04-28 14:20:01  
382           start  2017-04-28 02:52:29  
...             ...                  ...  
342679        start  2017-04-28 10:18:20  
342684        start  2017-04-28 12:44:21  
342685        start  2017-04-28 12:44:21  
342944        start  2017-04-28 07:07:17  
342945        start  2017-04-28 07:07:17  

[4124 rows x 5 columns]

In [9]:
# Drop duplicated rows
df1.drop_duplicates(keep='first')

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
343136    2443347  465784886_73090545671  136329900_10529659391_316129   
343137    2443347  465784886_73090545671  136329900_10529659391_316129   
343138    2443347  465784886_73090545671  136329900_10529659391_316129   
343139    2443347  465784886_73090545671  136329900_10529659391_316129   
343140    2443347  465784886_73090545671  136329900_10529659391_316129   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
343136      confirm  2017-03-31 15:15:46  
343137       step_3  2017-03-31 15:14:53  
343138       step_2  2017-03-31 15:12:08  
343139       step_1  2017-03-31 15:11:37  
343140        start  2017-03-31 15:10:47  

[341046 rows x 5 columns]

In [10]:
df1.describe(include='object')

visitor_id                     visit_id process_step  \
count                  343141                       343141       343141   
unique                  62936                        75256            5   
top     699275239_82397698587  712824876_8175482950_365042        start   
freq                       66                           61       108910   

                  date_time  
count                343141  
unique               282666  
top     2017-03-29 11:24:54  
freq                     14

In [11]:
df2 = pd.read_csv(file4)


In [12]:
list(df2.columns)

['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']

In [13]:
df2.shape

(412264, 5)

In [14]:
df2.dtypes

client_id        int64
visitor_id      object
visit_id        object
process_step    object
date_time       object
dtype: object

In [15]:
df2.nunique()

client_id        67430
visitor_id       71042
visit_id         82841
process_step         5
date_time       346697
dtype: int64

In [16]:
df2.describe(include='object')

visitor_id                      visit_id process_step  \
count                 412264                        412264       412264   
unique                 71042                         82841            5   
top     722943003_3441581446  875138661_34710212496_881092        start   
freq                     104                           104       135035   

                  date_time  
count                412264  
unique               346697  
top     2017-05-02 10:53:01  
freq                     24

In [17]:
df2.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
412259    False
412260    False
412261    False
412262    False
412263    False
Length: 412264, dtype: bool

In [18]:
df3 = pd.read_csv(file2)


In [19]:
list(df3.columns)

['client_id', 'Variation']

In [20]:
df3.shape

(70609, 2)

In [21]:
df3.dtypes

client_id     int64
Variation    object
dtype: object

In [22]:
df3.describe(include='object')

Variation
count      50500
unique         2
top         Test
freq       26968

# we have 70609 rows but count 50500. we have  20109 missing values

In [23]:
df3.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
70604    False
70605    False
70606    False
70607    False
70608    False
Length: 70609, dtype: bool

In [24]:
df4 = pd.read_csv(file1)
df4

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

In [25]:
df4.columns

Index(['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr',
       'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth'],
      dtype='object')

In [26]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    70595 non-null  float64
 2   clnt_tenure_mnth  70595 non-null  float64
 3   clnt_age          70594 non-null  float64
 4   gendr             70595 non-null  object 
 5   num_accts         70595 non-null  float64
 6   bal               70595 non-null  float64
 7   calls_6_mnth      70595 non-null  float64
 8   logons_6_mnth     70595 non-null  float64
dtypes: float64(7), int64(1), object(1)
memory usage: 4.8+ MB


In [27]:
df4.shape

(70609, 9)

In [28]:
df4.dtypes

client_id             int64
clnt_tenure_yr      float64
clnt_tenure_mnth    float64
clnt_age            float64
gendr                object
num_accts           float64
bal                 float64
calls_6_mnth        float64
logons_6_mnth       float64
dtype: object

In [29]:
df4.describe()

client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  7.060900e+04    70595.000000      70595.000000  70594.000000   
mean   5.004992e+06       12.052950        150.659367     46.442240   
std    2.877278e+06        6.871819         82.089854     15.591273   
min    1.690000e+02        2.000000         33.000000     13.500000   
25%    2.519329e+06        6.000000         82.000000     32.500000   
50%    5.016978e+06       11.000000        136.000000     47.000000   
75%    7.483085e+06       16.000000        192.000000     59.000000   
max    9.999839e+06       62.000000        749.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  70595.000000  7.059500e+04  70595.000000   70595.000000  
mean       2.255528  1.474452e+05      3.382478       5.566740  
std        0.534997  3.015087e+05      2.236580       2.353286  
min        1.000000  1.378942e+04      0.000000       1.000000  
25%        2.000000  3.734683e+04      1.000000       4.000000  
50%        2.000000  6.333290e+04      3.000000       5.000000  
75%        2.000000  1.375449e+05      6.000000       7.000000  
max        8.000000  1.632004e+07      7.000000       9.000000

In [30]:
df4.describe(include='object')

gendr
count   70595
unique      4
top         U
freq    24122

In [31]:
summary_df = df4.describe(include='all')

# number of rows= 70609  'count'=70595  # ~14 millions , ! clnt_age 

In [32]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    70595 non-null  float64
 2   clnt_tenure_mnth  70595 non-null  float64
 3   clnt_age          70594 non-null  float64
 4   gendr             70595 non-null  object 
 5   num_accts         70595 non-null  float64
 6   bal               70595 non-null  float64
 7   calls_6_mnth      70595 non-null  float64
 8   logons_6_mnth     70595 non-null  float64
dtypes: float64(7), int64(1), object(1)
memory usage: 4.8+ MB


In [33]:
df4.nunique()

client_id           70609
clnt_tenure_yr         54
clnt_tenure_mnth      482
clnt_age              165
gendr                   4
num_accts               8
bal                 70328
calls_6_mnth            8
logons_6_mnth           9
dtype: int64

In [34]:
df4['gendr'].nunique()

4

In [35]:
df4['gendr'].unique()

array(['U', 'M', 'F', nan, 'X'], dtype=object)

In [36]:
df4.select_dtypes(include='number')

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age  num_accts  \
0         836976             6.0              73.0      60.5        2.0   
1        2304905             7.0              94.0      58.0        2.0   
2        1439522             5.0              64.0      32.0        2.0   
3        1562045            16.0             198.0      49.0        2.0   
4        5126305            12.0             145.0      33.0        2.0   
...          ...             ...               ...       ...        ...   
70604    7993686             4.0              56.0      38.5        3.0   
70605    8981690            12.0             148.0      31.0        2.0   
70606     333913            16.0             198.0      61.5        2.0   
70607    1573142            21.0             255.0      68.0        3.0   
70608    5602139            21.0             254.0      59.5        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 8 columns]

In [37]:
df4.iloc[:,:]

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

In [38]:
df4.dtypes

client_id             int64
clnt_tenure_yr      float64
clnt_tenure_mnth    float64
clnt_age            float64
gendr                object
num_accts           float64
bal                 float64
calls_6_mnth        float64
logons_6_mnth       float64
dtype: object

In [39]:
# merging the dataframes from the df_final_web_data_pt_1.txt' and df_final_web_data_pt_2.txt'
dfs = pd.concat([df1,df2], ignore_index=True)
dfs

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

In [40]:
# Display duplicated rows 
dfs.loc[dfs.duplicated(keep=False)]

client_id             visitor_id                      visit_id  \
364       9358626  148475529_63668223960  223297395_36250329195_832161   
365       9358626  148475529_63668223960  223297395_36250329195_832161   
367       5376894  622509328_44408201172  688984457_43441834354_912755   
368       5376894  622509328_44408201172  688984457_43441834354_912755   
382        438567  923318467_50708963627  330543236_29863358529_771432   
...           ...                    ...                           ...   
755306    4408245   14058946_73585431797  994148957_84373994761_991865   
755307    4408245   14058946_73585431797  994148957_84373994761_991865   
755308    4408245   14058946_73585431797  994148957_84373994761_991865   
755357    3897324   80177429_24869309091  149796456_64325617570_112561   
755358    3897324   80177429_24869309091  149796456_64325617570_112561   

       process_step            date_time  
364           start  2017-04-28 12:27:28  
365           start  2017-04-28 12:27:28  
367           start  2017-04-28 14:20:01  
368           start  2017-04-28 14:20:01  
382           start  2017-04-28 02:52:29  
...             ...                  ...  
755306        start  2017-06-06 09:09:44  
755307        start  2017-06-06 08:51:16  
755308        start  2017-06-06 08:51:16  
755357        start  2017-05-04 08:21:30  
755358        start  2017-05-04 08:21:30  

[19816 rows x 5 columns]

In [41]:
# Drop duplicated rows
dfs = dfs.drop_duplicates(keep='first')
dfs

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[744641 rows x 5 columns]

In [42]:
# convert the column date-time to datetime type
dfs["date_time"] = pd.to_datetime(dfs["date_time"])
dfs["date_time"]

C:\Users\davyg\AppData\Local\Temp\ipykernel_25256\2910821914.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfs["date_time"] = pd.to_datetime(dfs["date_time"])


0        2017-04-17 15:27:07
1        2017-04-17 15:26:51
2        2017-04-17 15:19:22
3        2017-04-17 15:19:13
4        2017-04-17 15:18:04
                 ...        
755400   2017-05-24 18:46:10
755401   2017-05-24 18:45:29
755402   2017-05-24 18:44:51
755403   2017-05-24 18:44:34
755404   2017-06-03 12:17:09
Name: date_time, Length: 744641, dtype: datetime64[ns]

In [43]:
dfs.dtypes

client_id                int64
visitor_id              object
visit_id                object
process_step            object
date_time       datetime64[ns]
dtype: object

In [44]:
# rename the column Variation
df3.rename(columns={'Variation':'variation'}, inplace=True)

In [46]:
df4.columns

Index(['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr',
       'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth'],
      dtype='object')

In [47]:
# Display the list of the clients who were part of the experiment
df3.isna().sum()*100/len(df3)

client_id     0.000000
variation    28.479372
dtype: float64

In [48]:
# Display the list of the clients who were part of the experiment
df3['variation'].value_counts()

variation
Test       26968
Control    23532
Name: count, dtype: int64

In [49]:
#merfe dfs + experiment(df3) to have Variaton = Control OR test)
dfs_with_variation = pd.merge(dfs, df3, on="client_id", how="left")
dfs_with_variation

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
744636    9668240   388766751_9038881013   922267647_3096648104_968866   
744637    9668240   388766751_9038881013   922267647_3096648104_968866   
744638    9668240   388766751_9038881013   922267647_3096648104_968866   
744639    9668240   388766751_9038881013   922267647_3096648104_968866   
744640     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step           date_time variation  
0            step_3 2017-04-17 15:27:07      Test  
1            step_2 2017-04-17 15:26:51      Test  
2            step_3 2017-04-17 15:19:22      Test  
3            step_2 2017-04-17 15:19:13      Test  
4            step_3 2017-04-17 15:18:04      Test  
...             ...                 ...       ...  
744636        start 2017-05-24 18:46:10       NaN  
744637        start 2017-05-24 18:45:29       NaN  
744638       step_1 2017-05-24 18:44:51       NaN  
744639        start 2017-05-24 18:44:34       NaN  
744640        start 2017-06-03 12:17:09       NaN  

[744641 rows x 6 columns]

In [50]:
dfs_with_variation[dfs_with_variation['variation'] == 'Test']

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
629299    1574008  117364417_77840596075  528720790_71583064618_169151   
629300    1574008  117364417_77840596075  528720790_71583064618_169151   
629301    1574008  117364417_77840596075  528720790_71583064618_169151   
629302    1574008  117364417_77840596075  528720790_71583064618_169151   
629303    1574008  117364417_77840596075  528720790_71583064618_169151   

       process_step           date_time variation  
0            step_3 2017-04-17 15:27:07      Test  
1            step_2 2017-04-17 15:26:51      Test  
2            step_3 2017-04-17 15:19:22      Test  
3            step_2 2017-04-17 15:19:13      Test  
4            step_3 2017-04-17 15:18:04      Test  
...             ...                 ...       ...  
629299      confirm 2017-05-06 23:51:05      Test  
629300       step_3 2017-05-06 23:50:40      Test  
629301       step_2 2017-05-06 23:43:47      Test  
629302       step_1 2017-05-06 23:43:34      Test  
629303        start 2017-05-06 23:43:27      Test  

[176699 rows x 6 columns]

In [51]:
# Display the % of na values for each column
dfs_with_variation.isna().sum()*100/len(dfs_with_variation)

client_id        0.000000
visitor_id       0.000000
visit_id         0.000000
process_step     0.000000
date_time        0.000000
variation       57.397592
dtype: float64

In [52]:
df4

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

In [53]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    70595 non-null  float64
 2   clnt_tenure_mnth  70595 non-null  float64
 3   clnt_age          70594 non-null  float64
 4   gendr             70595 non-null  object 
 5   num_accts         70595 non-null  float64
 6   bal               70595 non-null  float64
 7   calls_6_mnth      70595 non-null  float64
 8   logons_6_mnth     70595 non-null  float64
dtypes: float64(7), int64(1), object(1)
memory usage: 4.8+ MB


In [54]:
df4.isna().sum()*100/len(df4)

client_id           0.000000
clnt_tenure_yr      0.019828
clnt_tenure_mnth    0.019828
clnt_age            0.021244
gendr               0.019828
num_accts           0.019828
bal                 0.019828
calls_6_mnth        0.019828
logons_6_mnth       0.019828
dtype: float64

In [55]:
df4.gendr.unique()

array(['U', 'M', 'F', nan, 'X'], dtype=object)

In [56]:
df4[df4['gendr'] == 'U']

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
5        3727881             5.0              71.0      30.5     U        2.0   
6         272934             5.0              66.0      58.5     U        2.0   
...          ...             ...               ...       ...   ...        ...   
70593    7217125            14.0             178.0      58.5     U        2.0   
70598    7807844             5.0              70.0      37.0     U        2.0   
70599    4851159             6.0              79.0      60.0     U        2.0   
70603    2489338             5.0              66.0      17.5     U        2.0   
70604    7993686             4.0              56.0      38.5     U        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
5        23915.60           0.0            3.0  
6        27021.42           2.0            5.0  
...           ...           ...            ...  
70593   319382.04           5.0            6.0  
70598   493497.05           2.0            2.0  
70599    74501.76           4.0            4.0  
70603    17869.95           4.0            4.0  
70604  1411062.68           5.0            5.0  

[24122 rows x 9 columns]

In [57]:
df4[df4['gendr'] == 'U'].isna().sum()*100/len(df4['gendr'])

client_id           0.0
clnt_tenure_yr      0.0
clnt_tenure_mnth    0.0
clnt_age            0.0
gendr               0.0
num_accts           0.0
bal                 0.0
calls_6_mnth        0.0
logons_6_mnth       0.0
dtype: float64

In [58]:
#merfe dfs_with_variation + client informations (df4) 
df_merge = pd.merge(dfs_with_variation, df4, on="client_id", how="left")
df_merge

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
744636    9668240   388766751_9038881013   922267647_3096648104_968866   
744637    9668240   388766751_9038881013   922267647_3096648104_968866   
744638    9668240   388766751_9038881013   922267647_3096648104_968866   
744639    9668240   388766751_9038881013   922267647_3096648104_968866   
744640     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step           date_time variation  clnt_tenure_yr  \
0            step_3 2017-04-17 15:27:07      Test             5.0   
1            step_2 2017-04-17 15:26:51      Test             5.0   
2            step_3 2017-04-17 15:19:22      Test             5.0   
3            step_2 2017-04-17 15:19:13      Test             5.0   
4            step_3 2017-04-17 15:18:04      Test             5.0   
...             ...                 ...       ...             ...   
744636        start 2017-05-24 18:46:10       NaN             NaN   
744637        start 2017-05-24 18:45:29       NaN             NaN   
744638       step_1 2017-05-24 18:44:51       NaN             NaN   
744639        start 2017-05-24 18:44:34       NaN             NaN   
744640        start 2017-06-03 12:17:09       NaN             NaN   

        clnt_tenure_mnth  clnt_age gendr  num_accts        bal  calls_6_mnth  \
0                   64.0      79.0     U        2.0  189023.86           1.0   
1                   64.0      79.0     U        2.0  189023.86           1.0   
2                   64.0      79.0     U        2.0  189023.86           1.0   
3                   64.0      79.0     U        2.0  189023.86           1.0   
4                   64.0      79.0     U        2.0  189023.86           1.0   
...                  ...       ...   ...        ...        ...           ...   
744636               NaN       NaN   NaN        NaN        NaN           NaN   
744637               NaN       NaN   NaN        NaN        NaN           NaN   
744638               NaN       NaN   NaN        NaN        NaN           NaN   
744639               NaN       NaN   NaN        NaN        NaN           NaN   
744640               NaN       NaN   NaN        NaN        NaN           NaN   

        logons_6_mnth  
0                 4.0  
1                 4.0  
2                 4.0  
3                 4.0  
4                 4.0  
...               ...  
744636            NaN  
744637            NaN  
744638            NaN  
744639            NaN  
744640            NaN  

[744641 rows x 14 columns]

'less than 40'

In [69]:
df = df_merge.copy()
df

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
744636    9668240   388766751_9038881013   922267647_3096648104_968866   
744637    9668240   388766751_9038881013   922267647_3096648104_968866   
744638    9668240   388766751_9038881013   922267647_3096648104_968866   
744639    9668240   388766751_9038881013   922267647_3096648104_968866   
744640     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step           date_time variation  clnt_tenure_yr  \
0            step_3 2017-04-17 15:27:07      Test             5.0   
1            step_2 2017-04-17 15:26:51      Test             5.0   
2            step_3 2017-04-17 15:19:22      Test             5.0   
3            step_2 2017-04-17 15:19:13      Test             5.0   
4            step_3 2017-04-17 15:18:04      Test             5.0   
...             ...                 ...       ...             ...   
744636        start 2017-05-24 18:46:10       NaN             NaN   
744637        start 2017-05-24 18:45:29       NaN             NaN   
744638       step_1 2017-05-24 18:44:51       NaN             NaN   
744639        start 2017-05-24 18:44:34       NaN             NaN   
744640        start 2017-06-03 12:17:09       NaN             NaN   

        clnt_tenure_mnth  clnt_age gendr  num_accts        bal  calls_6_mnth  \
0                   64.0      79.0     U        2.0  189023.86           1.0   
1                   64.0      79.0     U        2.0  189023.86           1.0   
2                   64.0      79.0     U        2.0  189023.86           1.0   
3                   64.0      79.0     U        2.0  189023.86           1.0   
4                   64.0      79.0     U        2.0  189023.86           1.0   
...                  ...       ...   ...        ...        ...           ...   
744636               NaN       NaN   NaN        NaN        NaN           NaN   
744637               NaN       NaN   NaN        NaN        NaN           NaN   
744638               NaN       NaN   NaN        NaN        NaN           NaN   
744639               NaN       NaN   NaN        NaN        NaN           NaN   
744640               NaN       NaN   NaN        NaN        NaN           NaN   

        logons_6_mnth  
0                 4.0  
1                 4.0  
2                 4.0  
3                 4.0  
4                 4.0  
...               ...  
744636            NaN  
744637            NaN  
744638            NaN  
744639            NaN  
744640            NaN  

[744641 rows x 14 columns]

In [ ]:
df.columns

In [81]:
df['process_step'].unique()

array(['start', 'step_1', 'step_2', 'step_3', 'confirm'], dtype=object)

In [70]:
# sorted values
df = df.sort_values(by=['client_id', 'visitor_id', 'visit_id', 'date_time'], ascending=[True, True, True, True])

In [ ]:
#df['gendr'].unique()

In [ ]:
# convert  X to U
#gender_mapping  = {'Unknow':'U', 'X':'U', 'F':'F', 'M':'M'}    #here male is replace by 0
#df['gendr'] = df['gendr'].map(gender_mapping)
#df['gendr']

In [ ]:
# fill the na values by U
#df['gendr'] = df['gendr'].fillna('U')
#df['gendr']

In [ ]:
#df['gendr'].unique()

In [73]:
# Display the list of the clients who were part of the experiment
#df['gendr'].value_counts()

gendr
M    149867
F    144036
Name: count, dtype: int64

In [72]:
# remove the rows without "F", "M"
df = df[df["gendr"].isin(["F", "M"])]
df

client_id             visitor_id                      visit_id  \
283823        169  201385055_71273495308  749567106_99161211863_557568   
283822        169  201385055_71273495308  749567106_99161211863_557568   
283821        169  201385055_71273495308  749567106_99161211863_557568   
283820        169  201385055_71273495308  749567106_99161211863_557568   
283819        169  201385055_71273495308  749567106_99161211863_557568   
...           ...                    ...                           ...   
272814    9999839  948322592_28990736246  715530422_68620416793_515645   
272813    9999839  948322592_28990736246  715530422_68620416793_515645   
272812    9999839  948322592_28990736246  715530422_68620416793_515645   
272811    9999839  948322592_28990736246  715530422_68620416793_515645   
272810    9999839  948322592_28990736246  715530422_68620416793_515645   

       process_step           date_time variation  clnt_tenure_yr  \
283823        start 2017-04-12 20:19:36       NaN            21.0   
283822       step_1 2017-04-12 20:19:45       NaN            21.0   
283821       step_2 2017-04-12 20:20:31       NaN            21.0   
283820       step_3 2017-04-12 20:22:05       NaN            21.0   
283819      confirm 2017-04-12 20:23:09       NaN            21.0   
...             ...                 ...       ...             ...   
272814        start 2017-03-29 12:11:32       NaN            13.0   
272813       step_1 2017-03-29 12:11:35       NaN            13.0   
272812       step_2 2017-03-29 12:11:50       NaN            13.0   
272811       step_3 2017-03-29 12:12:28       NaN            13.0   
272810      confirm 2017-03-29 12:13:03       NaN            13.0   

        clnt_tenure_mnth  clnt_age gendr  num_accts        bal  calls_6_mnth  \
283823             262.0      47.5     M        2.0  501570.72           4.0   
283822             262.0      47.5     M        2.0  501570.72           4.0   
283821             262.0      47.5     M        2.0  501570.72           4.0   
283820             262.0      47.5     M        2.0  501570.72           4.0   
283819             262.0      47.5     M        2.0  501570.72           4.0   
...                  ...       ...   ...        ...        ...           ...   
272814             160.0      28.5     F        2.0   67425.35           3.0   
272813             160.0      28.5     F        2.0   67425.35           3.0   
272812             160.0      28.5     F        2.0   67425.35           3.0   
272811             160.0      28.5     F        2.0   67425.35           3.0   
272810             160.0      28.5     F        2.0   67425.35           3.0   

        logons_6_mnth  
283823            4.0  
283822            4.0  
283821            4.0  
283820            4.0  
283819            4.0  
...               ...  
272814            3.0  
272813            3.0  
272812            3.0  
272811            3.0  
272810            3.0  

[293903 rows x 14 columns]

In [74]:
#remove rows without any variation
df = df[df["variation"].isin(["Control", "Test"])]
df

client_id             visitor_id                      visit_id  \
137375        647   66758770_53988066587   40369564_40101682850_311847   
137374        647   66758770_53988066587   40369564_40101682850_311847   
137373        647   66758770_53988066587   40369564_40101682850_311847   
137372        647   66758770_53988066587   40369564_40101682850_311847   
137369        647   66758770_53988066587   40369564_40101682850_311847   
...           ...                    ...                           ...   
106533    9999729  843385170_36953471821   493310979_9209676464_421146   
106532    9999729  843385170_36953471821   493310979_9209676464_421146   
106531    9999729  843385170_36953471821   493310979_9209676464_421146   
354472    9999832  145538019_54444341400  472154369_16714624241_585315   
354471    9999832  145538019_54444341400  472154369_16714624241_585315   

       process_step           date_time variation  clnt_tenure_yr  \
137375        start 2017-04-12 15:41:28      Test            12.0   
137374       step_1 2017-04-12 15:41:35      Test            12.0   
137373       step_2 2017-04-12 15:41:53      Test            12.0   
137372       step_3 2017-04-12 15:45:02      Test            12.0   
137369      confirm 2017-04-12 15:47:45      Test            12.0   
...             ...                 ...       ...             ...   
106533       step_1 2017-04-20 14:22:49      Test            10.0   
106532       step_2 2017-04-20 14:27:36      Test            10.0   
106531        start 2017-04-20 14:28:57      Test            10.0   
354472        start 2017-05-16 16:46:03      Test            23.0   
354471       step_1 2017-05-16 16:46:11      Test            23.0   

        clnt_tenure_mnth  clnt_age gendr  num_accts        bal  calls_6_mnth  \
137375             151.0      57.5     M        2.0   30525.80           0.0   
137374             151.0      57.5     M        2.0   30525.80           0.0   
137373             151.0      57.5     M        2.0   30525.80           0.0   
137372             151.0      57.5     M        2.0   30525.80           0.0   
137369             151.0      57.5     M        2.0   30525.80           0.0   
...                  ...       ...   ...        ...        ...           ...   
106533             124.0      31.0     F        3.0  107059.74           6.0   
106532             124.0      31.0     F        3.0  107059.74           6.0   
106531             124.0      31.0     F        3.0  107059.74           6.0   
354472             281.0      49.0     F        2.0  431887.61           1.0   
354471             281.0      49.0     F        2.0  431887.61           1.0   

        logons_6_mnth  
137375            4.0  
137374            4.0  
137373            4.0  
137372            4.0  
137369            4.0  
...               ...  
106533            9.0  
106532            9.0  
106531            9.0  
354472            4.0  
354471            4.0  

[209513 rows x 14 columns]

In [75]:
df["variation"].value_counts()

variation
Test       117457
Control     92056
Name: count, dtype: int64

In [76]:
df["gendr"].value_counts()

gendr
M    106703
F    102810
Name: count, dtype: int64

In [77]:
df["process_step"].value_counts()

process_step
start      67594
step_1     45209
step_2     36764
step_3     31420
confirm    28526
Name: count, dtype: int64

In [80]:
df_confirm = df[df["process_step"].isin(['confirm'])]
df_confirm

client_id             visitor_id                      visit_id  \
137369        647   66758770_53988066587   40369564_40101682850_311847   
13265        1195  766842522_69992551638  393817425_39015278493_996341   
584942       1336  920624746_32603333901  583743392_96265099036_939815   
584904       1336  920624746_32603333901  614001770_19101025926_112779   
584903       1336  920624746_32603333901  614001770_19101025926_112779   
...           ...                    ...                           ...   
113264    9996711  560523684_67875119133  200256953_19881035574_656660   
1312      9997114  708360389_95473182321     19940015_3197786491_51650   
118837    9998346  292425655_16607136645  189177304_69869411700_783154   
39450     9999009  212849256_28644888524  149965839_24823291142_245569   
467272    9999729  834634258_21862004160  870243567_56915814033_814203   

       process_step           date_time variation  clnt_tenure_yr  \
137369      confirm 2017-04-12 15:47:45      Test            12.0   
13265       confirm 2017-04-05 20:19:31   Control            21.0   
584942      confirm 2017-05-08 06:08:43      Test            48.0   
584904      confirm 2017-05-08 08:21:38      Test            48.0   
584903      confirm 2017-05-08 08:23:00      Test            48.0   
...             ...                 ...       ...             ...   
113264      confirm 2017-04-21 19:20:47   Control            11.0   
1312        confirm 2017-04-14 20:21:31   Control            22.0   
118837      confirm 2017-03-29 15:37:28   Control            50.0   
39450       confirm 2017-03-31 07:20:36      Test            18.0   
467272      confirm 2017-05-08 16:09:40      Test            10.0   

        clnt_tenure_mnth  clnt_age gendr  num_accts         bal  calls_6_mnth  \
137369             151.0      57.5     M        2.0    30525.80           0.0   
13265              262.0      54.5     M        2.0    28457.96           2.0   
584942             576.0      42.0     M        4.0   130537.18           6.0   
584904             576.0      42.0     M        4.0   130537.18           6.0   
584903             576.0      42.0     M        4.0   130537.18           6.0   
...                  ...       ...   ...        ...         ...           ...   
113264             142.0      42.5     M        2.0   122486.93           2.0   
1312               264.0      44.5     M        2.0    80740.80           1.0   
118837             602.0      61.5     F        2.0   149881.38           6.0   
39450              216.0      68.5     M        3.0  2055381.29           6.0   
467272             124.0      31.0     F        3.0   107059.74           6.0   

        logons_6_mnth  
137369            4.0  
13265             5.0  
584942            9.0  
584904            9.0  
584903            9.0  
...               ...  
113264            5.0  
1312              4.0  
118837            9.0  
39450             9.0  
467272            9.0  

[28526 rows x 14 columns]

In [83]:
# Create a age group
# Create the age group
def create_age_group(age):
    if age <= 40:
        return "less than 40"
    else:
        return "more than 40"
df['age_group'] = df['clnt_age'].apply(create_age_group)
df

C:\Users\davyg\AppData\Local\Temp\ipykernel_25256\875189680.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age_group'] = df['clnt_age'].apply(create_age_group)


client_id             visitor_id                      visit_id  \
137375        647   66758770_53988066587   40369564_40101682850_311847   
137374        647   66758770_53988066587   40369564_40101682850_311847   
137373        647   66758770_53988066587   40369564_40101682850_311847   
137372        647   66758770_53988066587   40369564_40101682850_311847   
137369        647   66758770_53988066587   40369564_40101682850_311847   
...           ...                    ...                           ...   
106533    9999729  843385170_36953471821   493310979_9209676464_421146   
106532    9999729  843385170_36953471821   493310979_9209676464_421146   
106531    9999729  843385170_36953471821   493310979_9209676464_421146   
354472    9999832  145538019_54444341400  472154369_16714624241_585315   
354471    9999832  145538019_54444341400  472154369_16714624241_585315   

       process_step           date_time variation  clnt_tenure_yr  \
137375        start 2017-04-12 15:41:28      Test            12.0   
137374       step_1 2017-04-12 15:41:35      Test            12.0   
137373       step_2 2017-04-12 15:41:53      Test            12.0   
137372       step_3 2017-04-12 15:45:02      Test            12.0   
137369      confirm 2017-04-12 15:47:45      Test            12.0   
...             ...                 ...       ...             ...   
106533       step_1 2017-04-20 14:22:49      Test            10.0   
106532       step_2 2017-04-20 14:27:36      Test            10.0   
106531        start 2017-04-20 14:28:57      Test            10.0   
354472        start 2017-05-16 16:46:03      Test            23.0   
354471       step_1 2017-05-16 16:46:11      Test            23.0   

        clnt_tenure_mnth  clnt_age gendr  num_accts        bal  calls_6_mnth  \
137375             151.0      57.5     M        2.0   30525.80           0.0   
137374             151.0      57.5     M        2.0   30525.80           0.0   
137373             151.0      57.5     M        2.0   30525.80           0.0   
137372             151.0      57.5     M        2.0   30525.80           0.0   
137369             151.0      57.5     M        2.0   30525.80           0.0   
...                  ...       ...   ...        ...        ...           ...   
106533             124.0      31.0     F        3.0  107059.74           6.0   
106532             124.0      31.0     F        3.0  107059.74           6.0   
106531             124.0      31.0     F        3.0  107059.74           6.0   
354472             281.0      49.0     F        2.0  431887.61           1.0   
354471             281.0      49.0     F        2.0  431887.61           1.0   

        logons_6_mnth     age_group  
137375            4.0  more than 40  
137374            4.0  more than 40  
137373            4.0  more than 40  
137372            4.0  more than 40  
137369            4.0  more than 40  
...               ...           ...  
106533            9.0  less than 40  
106532            9.0  less than 40  
106531            9.0  less than 40  
354472            4.0  more than 40  
354471            4.0  more than 40  

[209513 rows x 15 columns]

In [84]:
df['age_group'].value_counts()

age_group
more than 40    154162
less than 40     55351
Name: count, dtype: int64

In [86]:
df[['clnt_age', 'age_group']]

clnt_age     age_group
137375      57.5  more than 40
137374      57.5  more than 40
137373      57.5  more than 40
137372      57.5  more than 40
137369      57.5  more than 40
...          ...           ...
106533      31.0  less than 40
106532      31.0  less than 40
106531      31.0  less than 40
354472      49.0  more than 40
354471      49.0  more than 40

[209513 rows x 2 columns]

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 209513 entries, 137375 to 354471
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   client_id         209513 non-null  int64         
 1   visitor_id        209513 non-null  object        
 2   visit_id          209513 non-null  object        
 3   process_step      209513 non-null  object        
 4   date_time         209513 non-null  datetime64[ns]
 5   variation         209513 non-null  object        
 6   clnt_tenure_yr    209513 non-null  float64       
 7   clnt_tenure_mnth  209513 non-null  float64       
 8   clnt_age          209501 non-null  float64       
 9   gendr             209513 non-null  object        
 10  num_accts         209513 non-null  float64       
 11  bal               209513 non-null  float64       
 12  calls_6_mnth      209513 non-null  float64       
 13  logons_6_mnth     209513 non-null  float64       
 14  age_

In [87]:
# Move column from position 14 to position 9
def move_column(df, old_pos, new_pos): 
    cols = list(df.columns) 
    col = cols.pop(old_pos) 
    cols.insert(new_pos, col) 
    return df[cols] 
    
df = move_column(df, 14, 9)
df

client_id             visitor_id                      visit_id  \
137375        647   66758770_53988066587   40369564_40101682850_311847   
137374        647   66758770_53988066587   40369564_40101682850_311847   
137373        647   66758770_53988066587   40369564_40101682850_311847   
137372        647   66758770_53988066587   40369564_40101682850_311847   
137369        647   66758770_53988066587   40369564_40101682850_311847   
...           ...                    ...                           ...   
106533    9999729  843385170_36953471821   493310979_9209676464_421146   
106532    9999729  843385170_36953471821   493310979_9209676464_421146   
106531    9999729  843385170_36953471821   493310979_9209676464_421146   
354472    9999832  145538019_54444341400  472154369_16714624241_585315   
354471    9999832  145538019_54444341400  472154369_16714624241_585315   

       process_step           date_time variation  clnt_tenure_yr  \
137375        start 2017-04-12 15:41:28      Test            12.0   
137374       step_1 2017-04-12 15:41:35      Test            12.0   
137373       step_2 2017-04-12 15:41:53      Test            12.0   
137372       step_3 2017-04-12 15:45:02      Test            12.0   
137369      confirm 2017-04-12 15:47:45      Test            12.0   
...             ...                 ...       ...             ...   
106533       step_1 2017-04-20 14:22:49      Test            10.0   
106532       step_2 2017-04-20 14:27:36      Test            10.0   
106531        start 2017-04-20 14:28:57      Test            10.0   
354472        start 2017-05-16 16:46:03      Test            23.0   
354471       step_1 2017-05-16 16:46:11      Test            23.0   

        clnt_tenure_mnth  clnt_age     age_group gendr  num_accts        bal  \
137375             151.0      57.5  more than 40     M        2.0   30525.80   
137374             151.0      57.5  more than 40     M        2.0   30525.80   
137373             151.0      57.5  more than 40     M        2.0   30525.80   
137372             151.0      57.5  more than 40     M        2.0   30525.80   
137369             151.0      57.5  more than 40     M        2.0   30525.80   
...                  ...       ...           ...   ...        ...        ...   
106533             124.0      31.0  less than 40     F        3.0  107059.74   
106532             124.0      31.0  less than 40     F        3.0  107059.74   
106531             124.0      31.0  less than 40     F        3.0  107059.74   
354472             281.0      49.0  more than 40     F        2.0  431887.61   
354471             281.0      49.0  more than 40     F        2.0  431887.61   

        calls_6_mnth  logons_6_mnth  
137375           0.0            4.0  
137374           0.0            4.0  
137373           0.0            4.0  
137372           0.0            4.0  
137369           0.0            4.0  
...              ...            ...  
106533           6.0            9.0  
106532           6.0            9.0  
106531           6.0            9.0  
354472           1.0            4.0  
354471           1.0            4.0  

[209513 rows x 15 columns]

In [88]:
df.to_csv('df_clean_group_age.csv', index=False)

In [92]:
df["variation"].value_counts()

variation
Test       117457
Control     92056
Name: count, dtype: int64

In [94]:
# Create a table for Tests and another for Controls! 
df_control = df[df["variation"] == "Control"].copy()
df

client_id             visitor_id                      visit_id  \
137375        647   66758770_53988066587   40369564_40101682850_311847   
137374        647   66758770_53988066587   40369564_40101682850_311847   
137373        647   66758770_53988066587   40369564_40101682850_311847   
137372        647   66758770_53988066587   40369564_40101682850_311847   
137369        647   66758770_53988066587   40369564_40101682850_311847   
...           ...                    ...                           ...   
106533    9999729  843385170_36953471821   493310979_9209676464_421146   
106532    9999729  843385170_36953471821   493310979_9209676464_421146   
106531    9999729  843385170_36953471821   493310979_9209676464_421146   
354472    9999832  145538019_54444341400  472154369_16714624241_585315   
354471    9999832  145538019_54444341400  472154369_16714624241_585315   

       process_step           date_time variation  clnt_tenure_yr  \
137375        start 2017-04-12 15:41:28      Test            12.0   
137374       step_1 2017-04-12 15:41:35      Test            12.0   
137373       step_2 2017-04-12 15:41:53      Test            12.0   
137372       step_3 2017-04-12 15:45:02      Test            12.0   
137369      confirm 2017-04-12 15:47:45      Test            12.0   
...             ...                 ...       ...             ...   
106533       step_1 2017-04-20 14:22:49      Test            10.0   
106532       step_2 2017-04-20 14:27:36      Test            10.0   
106531        start 2017-04-20 14:28:57      Test            10.0   
354472        start 2017-05-16 16:46:03      Test            23.0   
354471       step_1 2017-05-16 16:46:11      Test            23.0   

        clnt_tenure_mnth  clnt_age     age_group gendr  num_accts        bal  \
137375             151.0      57.5  more than 40     M        2.0   30525.80   
137374             151.0      57.5  more than 40     M        2.0   30525.80   
137373             151.0      57.5  more than 40     M        2.0   30525.80   
137372             151.0      57.5  more than 40     M        2.0   30525.80   
137369             151.0      57.5  more than 40     M        2.0   30525.80   
...                  ...       ...           ...   ...        ...        ...   
106533             124.0      31.0  less than 40     F        3.0  107059.74   
106532             124.0      31.0  less than 40     F        3.0  107059.74   
106531             124.0      31.0  less than 40     F        3.0  107059.74   
354472             281.0      49.0  more than 40     F        2.0  431887.61   
354471             281.0      49.0  more than 40     F        2.0  431887.61   

        calls_6_mnth  logons_6_mnth  
137375           0.0            4.0  
137374           0.0            4.0  
137373           0.0            4.0  
137372           0.0            4.0  
137369           0.0            4.0  
...              ...            ...  
106533           6.0            9.0  
106532           6.0            9.0  
106531           6.0            9.0  
354472           1.0            4.0  
354471           1.0            4.0  

[209513 rows x 15 columns]

In [ ]:
# Create a table for Tests and another for Controls! 
df_test = df[df["variation"] == "Test"].copy()
df_test

In [99]:
df[df['visitor_id']== '145538019_54444341400']

client_id             visitor_id                      visit_id  \
354472    9999832  145538019_54444341400  472154369_16714624241_585315   
354471    9999832  145538019_54444341400  472154369_16714624241_585315   

       process_step           date_time variation  clnt_tenure_yr  \
354472        start 2017-05-16 16:46:03      Test            23.0   
354471       step_1 2017-05-16 16:46:11      Test            23.0   

        clnt_tenure_mnth  clnt_age     age_group gendr  num_accts        bal  \
354472             281.0      49.0  more than 40     F        2.0  431887.61   
354471             281.0      49.0  more than 40     F        2.0  431887.61   

        calls_6_mnth  logons_6_mnth  
354472           1.0            4.0  
354471           1.0            4.0

In [ ]:

# Clients with higher tenure years are likely to have higher average balances than more recent clients? Inverse? non influence?Davy

# The number of accounts per client may differ by gender? Davy

# The probability that the client reached the final confirmation page for test and control variation

In [104]:
# Filter only rows where the client reached the 'confirm' step 
df_confirm = df[df["process_step"].isin(['confirm'])]
df_confirm

# Compute probability for each variation 
probabilities = df_confirm.groupby('variation').size() / df.groupby('variation').size() 

print(probabilities)

variation
Control    0.124413
Test       0.145355
dtype: float64


In [105]:
df.columns

Index(['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time',
       'variation', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age',
       'age_group', 'gendr', 'num_accts', 'bal', 'calls_6_mnth',
       'logons_6_mnth'],
      dtype='object')

In [ ]:
df4["bal"].value_counts().unique()

# Clients in the test variation are more likely to reach the final confirmation step than clients in the control variation. (does the tes variation improve the conversion funnel?)

In [107]:
# Contingency table 
table = pd.crosstab(df['variation'], df_confirm) 
chi2, p, dof, expected = stats.chi2_contingency(table) 
print("Chi-square:", chi2) 
print("p-value:", p) 
print("Expected frequencies:\n", expected)

ValueError: 2

# The test variation increases engagement (more steps completed) among younger clients compared to older clients.

In [109]:
# Number of steps by visit
engagement = (
    df
    .groupby(['visit_id', 'variation', 'age_group'])
    ['process_step']
    .nunique()
    .reset_index(name='n_steps')
)

engagement

visit_id variation     age_group  n_steps
0       100019538_17884295066_43909      Test  less than 40        5
1      100022086_87870757897_149620      Test  more than 40        5
2      100030127_47967100085_936361   Control  less than 40        1
3      100037962_47432393712_705583   Control  more than 40        2
4       10006594_66157970412_679648   Control  more than 40        5
...                             ...       ...           ...      ...
46329  999960019_60838685252_926860      Test  more than 40        3
46330  999971096_28827267783_236076      Test  less than 40        5
46331  999976049_95772503197_182554      Test  more than 40        5
46332  999984454_18731538378_781808      Test  more than 40        5
46333  999988789_76411676596_272843      Test  more than 40        3

[46334 rows x 4 columns]

In [116]:
engagement['age_binary'] = engagement['age_group'].map({
    'less than 40': '0-40',
    'more than 40': '40+'
})
engagement['age_binary']

0        0-40
1         40+
2        0-40
3         40+
4         40+
         ... 
46329     40+
46330    0-40
46331     40+
46332     40+
46333     40+
Name: age_binary, Length: 46334, dtype: object

In [117]:
import statsmodels.formula.api as smf

model = smf.ols(
    formula="""
    n_steps ~ C(variation)
             * C(age_binary)
             + clnt_tenure_yr
             + num_accts
             + bal
             + logons_6_mnth
    """,
    data=engagement
).fit()

model.summary()

ImportError: cannot import name '_lazywhere' from 'scipy._lib._util' (C:\Users\davyg\anaconda3\Lib\site-packages\scipy\_lib\_util.py)

# The impact of the test variation on reaching the confirmation step differs between long‑tenure and short‑tenure clients

# Clients in the test variation have a higher probability of progressing further in the process_step funnel (e.g., confirm) than those in the control variation, especially among clients with high logon frequency.

In [ ]:
voici mes colonnes "Index(['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time',
       'variation', 'clnt_tenure_yr'(how long the client has been with Vanguard in year), 'clnt_tenure_mnth', 'clnt_age',
       'age_group', 'gendr', 'num_accts', 'bal', 'calls_6_mnth'(record the number of the time the client reched out over a call in the past 6 months),
       'logons_6_mnth'],"(is the frequency with wich the client logged onto the plaform Vanguard over last 6 months) . il s'agit d'un fichier avec des données permettant d'analyser les comportements et préferences des clients.
1.écris les H0 / H1 pour ' The test variation increases engagement (more steps completed) among younger clients compared to older clients.'

2.proposer les tests statistiques adaptés (chi‑square, logistic regression, ANOVA…),

3.écris le code pandas/statsmodels pour tester chaque hypothèse,

4.aide_moi à construire un funnel complet test vs control.

In [ ]:
# H0 : There is no significant relationship between client tenure and average account balances
# with significance level = 0.05

from scipy.stats import pearsonr, spearmanr

# Step 1: Select relevant columns and drop the missing values
df_tenure_balance = df[['clnt_tenure_yr', 'bal']].dropna()

# Step 2: Pearson correlation test
pearson_corr, pearson_p = pearsonr(df_tenure_balance['clnt_tenure_yr'], df_tenure_balance['bal'])

print("Pearson correlation:", pearson_corr)
print("p-value:", pearson_p)



p<0 there is a significant relationship between the both variables. So we reject the H0

#### Clients with longer tenure at Vanguard tend to have higher average account balances compared to newer clients
H0 : There is no significant difference between the average account balances of clients with longer tenure compared to newer clients

H1 : There is significant difference between the average account balances of clients with longer tenure compared to newer clients
we will considere that a client with clnt_tenure_yr <= 5 is a new client
alpha = 0.05

In [ ]:
from scipy.stats import ttest_ind

# Select relevant columns and drop the missing values
df_tenure_balance = df4[['clnt_tenure_yr', 'bal']].dropna()

# Split the dataset into newer and longer tenuregroups
long_clients_df = df_tenure_balance[df_tenure_balance['clnt_tenure_yr'] > 5]['bal']
new_clients_df = df_tenure_balance[df_tenure_balance['clnt_tenure_yr'] <= 5]['bal']


# Perform statistics 
mean_long = long_clients_df.mean()
std_long = long_clients_df.std(ddof=1)
mean_new = new_clients_df.mean()
std_new = new_clients_df.std(ddof=1)


#perform independent t-test
t_stat, p_value = ttest_ind(long_clients_df, new_clients_df, equal_var=False)

# Print results
print(f"Clients with longer tenure: mean={mean_long:.2f}, std={std_long:.2f}")
print(f"Newer clients: mean={mean_new:.2f}, std={std_new:.2f}")
print(f"T-statistic={t_stat:.2f}, p-value={p_value:.5f}")


Longer-tenure clients have significantly higher average balances with the mean of the longer tenure almost the doble of the new
p-value < 0.05 => we reject the H0. and conclude that There is significant difference between the average account balances of clients with longer tenure compared to newer client.

In [ ]:
# Create a new column for grouping
avg_balance = df_tenure_balance.groupby('tenure_group')['bal'].mean()

plt.figure(figsize=(6,5))
avg_balance.plot(kind='bar', color=['skyblue','orange'])
plt.title("Average Account Balance by Tenure Group")
plt.xlabel("Client Tenure Group")
plt.ylabel("Average Balance")
plt.show()

In [ ]:
df.columns

In [ ]:
df['gendr'].unique()

In [ ]:
df['client_id'].nunique()

In [ ]:
df['visit_id'].nunique()

In [ ]:
df['gendr'].value_counts()

In [ ]:
df4['gendr'].value_counts()

In [ ]:
df1